CI.hs

```diff
@@ -217,26 +224,53 @@ buildDists
+    if noGhcCheckout then do
+      cmd "cd ghc && git clean -xdf && git submodule foreach git clean -xdf && git submodule foreach git checkout . && git checkout ."
+      case ghcFlavor of
+          Ghc921 -> cmd "cd ghc && git checkout ghc-9.2.1-alpha2"
+          Ghc901 -> cmd "cd ghc && git checkout ghc-9.0.1-release"
+          Ghc8101 -> cmd "cd ghc && git checkout ghc-8.10.1-release"
+          Ghc8102 -> cmd "cd ghc && git checkout ghc-8.10.2-release"
+          Ghc8103 -> cmd "cd ghc && git checkout ghc-8.10.3-release"
+          Ghc8104 -> cmd "cd ghc && git checkout ghc-8.10.4-release"
+          Ghc881 -> cmd "cd ghc && git checkout ghc-8.8.1-release"
+          Ghc882 -> cmd "cd ghc && git checkout ghc-8.8.2-release"
+          Ghc883 -> cmd "cd ghc && git checkout ghc-8.8.3-release"
+          Ghc884 -> cmd "cd ghc && git checkout ghc-8.8.4-release"
+          Da { mergeBaseSha, patches, upstream } -> do
+              cmd $ "cd ghc && git checkout " <> mergeBaseSha
+              -- Apply Digital Asset extensions.
+              cmd $ "cd ghc && git remote add upstream " <> upstream
+              cmd "cd ghc && git fetch upstream"
+              cmd $ "cd ghc && git -c user.name=\"Cookie Monster\" -c user.email=cookie.monster@seasame-street.com merge --no-edit " <> unwords patches
+          GhcMaster hash -> cmd $ "cd ghc && git checkout " ++ hash
+     else do
+        -- Get a clone of ghc.
+        cmd "git clone https://gitlab.haskell.org/ghc/ghc.git"
+        case ghcFlavor of
+            Ghc921 -> cmd "cd ghc && git fetch --tags && git checkout ghc-9.2.1-alpha2"
+            Ghc901 -> cmd "cd ghc && git fetch --tags && git checkout ghc-9.0.1-release"
+            Ghc8101 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.10.1-release"
+            Ghc8102 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.10.2-release"
+            Ghc8103 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.10.3-release"
+            Ghc8104 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.10.4-release"
+            Ghc881 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.8.1-release"
+            Ghc882 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.8.2-release"
+            Ghc883 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.8.3-release"
+            Ghc884 -> cmd "cd ghc && git fetch --tags && git checkout ghc-8.8.4-release"
+            Da { mergeBaseSha, patches, upstream } -> do
+                cmd $ "cd ghc && git fetch --tags && git checkout " <> mergeBaseSha
+                -- Apply Digital Asset extensions.
+                cmd $ "cd ghc && git remote add upstream " <> upstream
+                cmd "cd ghc && git fetch upstream"
+                cmd $ "cd ghc && git -c user.name=\"Cookie Monster\" -c user.email=cookie.monster@seasame-street.com merge --no-edit " <> unwords patches
+            GhcMaster hash -> cmd $ "cd ghc && git checkout " ++ hash
```

Comment:
This is duplicating 95% of the logic but the only difference is that in one case we call `git clone` (and for some reason `git fetch --tags`?) Can we share the 95% by calling the `git clone` step separately beforehand?

Reply:
absolutely. done.

Reply:
mind giving this an `/azp run` please?

Reply:
oops. introduced an error. sorry. fixed now.